File: go/src/RumbleRacing/core/file/track.go
package file

import (
	"fmt"
	"io"
	"log"
	"os"
	"rumble-reader/asset"
	"rumble-reader/asset/o3d"
	"rumble-reader/asset/txf"
	"rumble-reader/chunk"
	"rumble-reader/chunk/shoc"
	"strings"
)

type TrackFile struct {
	FileName       string
	FileSize       int64
	TopLevelChunks []chunk.TopLevelChunk
}

func ReadTrackChunks(file io.ReadSeeker) []chunk.TopLevelChunk {

	var chunks []chunk.TopLevelChunk

	var chunkIndex uint32 = 0
	for {
		pos, _ := file.Seek(0, io.SeekCurrent)
		chunkObj, err := readTopLevelChunk(file, chunkIndex)
		if err == io.EOF {
			// fmt.Println("reached end of file!")
			break
		}
		if err == io.ErrUnexpectedEOF {
			fmt.Println("Unexpected EOF — incomplete chunk at end of file.")
			break
		}
		if err != nil {
			log.Fatalf("Error reading chunk at 0x%X: %v", pos, err)
		}

		// Do not append empty FILL chunks, who cares about them.
		_, ok := chunkObj.(*chunk.Fill)

		if !ok {
			chunks = append(chunks, chunkObj)
			// only increment chunk index if we're actually adding a chunk
			chunkIndex++
		}
	}

	return chunks
}

func ReadTrackFile(filename string) TrackFile {
	file, err := os.Open(filename)

	if err != nil {
		log.Fatalf("Failed to open file: %v", err)
	}

	defer file.Close()

	info, err := file.Stat()

	if err != nil {
		log.Fatalf("Failed to get file info: %v", err)
	}

	// fmt.Printf("File: %s\nSize: %d bytes\n\n", info.Name(), info.Size())

	chunks := ReadTrackChunks(file)

	return TrackFile{
		FileName:       info.Name(),
		FileSize:       info.Size(),
		TopLevelChunks: chunks,
	}
}

func (t TrackFile) GetResourceList() *asset.RLst {

	// FE2 has multiple resource lists.
	// Let's just combine them and treat it as one giant list for now.
	var newList = asset.RLst{
		Count:   0,
		Entries: []asset.ResourceEntry{},
	}

	headers := t.getHeadersForType("RLst")

	for _, header := range headers {
		// fmt.Println(header.Unk0, header.AssetType, header.AssetIndex, header.TotalDataSize)

		rList, err := asset.ParseRLst(t.getDataForHeader(header), t.FileName)
		if err != nil {
			panic(err)
		}

		newList.Count += rList.Count
		newList.Entries = append(newList.Entries, rList.Entries...)
	}

	return &newList
}

func (t TrackFile) getHeadersForType(assetType string) []shoc.SHDR {
	var headers []shoc.SHDR

	// Get the SHDRs for all assetTypes in the file
	for _, chunk := range t.TopLevelChunks {
		// get SHOC
		shc, ok := chunk.(*shoc.Shoc)
		if ok {
			// Get Headers
			header, ok := shc.MetaData.(*shoc.SHDR)
			if ok {
				if header.AssetType == assetType {
					headers = append(headers, *header)
				}
			}
		}
	}

	return headers
}

func (t TrackFile) getHeaderForResource(res asset.ResourceEntry) *shoc.SHDR {
	for _, chunk := range t.TopLevelChunks {
		// get SHOC
		shc, ok := chunk.(*shoc.Shoc)
		if ok {
			// Get Headers
			header, ok := shc.MetaData.(*shoc.SHDR)
			if ok {
				if header.AssetType == res.TypeTag {
					// fmt.Println("target", res.TypeTag, res.ResourceIndex, "comparing", header.AssetType, header.AssetIndex)
					if header.AssetIndex == res.ResourceIndex {
						// fmt.Println("FOUND!", "target", res.TypeTag, res.ResourceIndex, "comparing", header.AssetType, header.AssetIndex)
						return header
					}
				}
			}
		}
	}

	return nil
}

func (t TrackFile) getDataForHeader(header shoc.SHDR) []byte {

	var assetData []byte

	// hdrShoc := t.TopLevelChunks[header.ShocIndex]
	// fmt.Println("getting data for:", header.AssetType, "size:", header.TotalDataSize, "asset idx:", header.AssetIndex, "| header address:", hdrShoc.StartAddress())
	// fmt.Println("Header addr", hdrShoc.StartAddress())

	shocCount := 1
	for {
		topLevel := t.TopLevelChunks[header.ShocIndex+uint32(shocCount)]

		theShoc, ok := topLevel.(*shoc.Shoc)

		if !ok {
			// skip past filler/unrelated chunks
			continue
		}

		// fmt.Println(theShoc.StartAddress(), theShoc.MetaData.FourCC(), "size:", len(theShoc.Data()))

		switch data := theShoc.MetaData.(type) {
		case *shoc.SDAT:
			assetData = append(assetData, data.Data()...)
		case *shoc.Rdat:
			decompressed, err := shoc.Decompress(data.Data(), int(data.OutBufferSize))
			if err != nil {
				fmt.Println("Error decompressing", header.AssetType, header.AssetIndex)
				return make([]byte, 0)
				// panic(err)
			}
			assetData = append(assetData, decompressed...)
		default:
			panic("Unhandled SHOC type!" + data.FourCC())
		}

		shocCount++
		// fmt.Println("total size", len(assetData))

		// making a pretty big assumption here,
		// that the contiguous shoc data equals header's size value
		if len(assetData) >= int(header.TotalDataSize) {
			break
		}
	}

	return assetData
}

func (t TrackFile) GetResource(resource asset.ResourceEntry) (asset.Asset, error) {
	// fmt.Println("attempting to get", resource.ResourceName, resource.TypeTag, "at resource index", resource.ResourceIndex)
	header := t.getHeaderForResource(resource)
	data := t.getDataForHeader(*header)
	switch resource.TypeTag {
	case "TxtR":
		return asset.ParseTxtR(data, *header)
	case "Cact":
		return asset.ParseActor(data)
	case "obf ":
		return o3d.ParseObf(data)
	case "o3d ":
		return o3d.ParseO3D(false, data, *header, resource.ResourceName)
	case "o3da":
		return o3d.ParseO3D(true, data, *header, resource.ResourceName)
	case "txf ", "txf2":
		name := fmt.Sprintf("%d_%s", resource.ResourceIndex, resource.ResourceName)
		return txf.ParseTXF(data, *header, name)
	default:
		return asset.ParseGenericAsset(data, strings.TrimSpace(resource.TypeTag), *header)
	}
}

File: src/RumbleRacing/file/track.ts
import ArrayBufferSlice from "../../ArrayBufferSlice";
import { TopLevelChunk, readTopLevelChunk } from "../chunk/chunk";
import { SHDR } from "../chunk/shoc/shdr";
import { decompress } from "../chunk/shoc/decompress";
import { parseRLst, RLst, ResourceEntry } from "../asset/RLst";
import { parseActor, Actor } from "../asset/Cact";
import { parseGenericAsset, GenericAsset } from "../asset/asset";
import { parseObf, Obf } from "../asset/o3d/obf";
import { parseO3D, O3D } from "../asset/o3d/o3d";
import { parseTXF, TXF } from "../asset/txf/TXF";
import { Gmd, parseGmd } from "../asset/gmd";

export type ParsedAsset = Actor | Obf | O3D | TXF | Gmd | GenericAsset;

export interface TrackFile {
  fileName: string;
  fileSize: number;
  topLevelChunks: TopLevelChunk[];
}

export function readTrackChunks(data: ArrayBufferSlice): TopLevelChunk[] {
  const view = data.createDataView();
  const cursor = { pos: 0 };
  const chunks: TopLevelChunk[] = [];
  let chunkIndex = 0;

  while (cursor.pos < data.byteLength) {
    try {
      const chunk = readTopLevelChunk(data, view, cursor, chunkIndex);
      if (chunk === null) break;
      if (chunk.kind === "FILL") continue;
      chunks.push(chunk);
      chunkIndex++;
    } catch (e) {
      if (cursor.pos >= data.byteLength) break;
      console.warn("Unexpected error reading chunk:", e);
      break;
    }
  }

  return chunks;
}

export function parseTrackFile(
  data: ArrayBufferSlice,
  fileName: string = "unknown",
): TrackFile {
  return {
    fileName,
    fileSize: data.byteLength,
    topLevelChunks: readTrackChunks(data),
  };
}

function getHeadersForType(track: TrackFile, assetType: string): SHDR[] {
  const headers: SHDR[] = [];
  for (const chunk of track.topLevelChunks) {
    if (chunk.kind !== "SHOC") continue;
    const meta = chunk.metadata;
    if (meta.kind !== "SHDR") continue;
    if (meta.assetType === assetType) headers.push(meta);
  }
  return headers;
}

function getHeaderForResource(
  track: TrackFile,
  res: ResourceEntry,
): SHDR | null {
  for (const chunk of track.topLevelChunks) {
    if (chunk.kind !== "SHOC") continue;
    const meta = chunk.metadata;
    if (meta.kind !== "SHDR") continue;
    if (
      meta.assetType === res.typeTag &&
      meta.assetIndex === res.resourceIndex
    ) {
      return meta;
    }
  }
  return null;
}

function getDataForHeader(track: TrackFile, header: SHDR): ArrayBufferSlice {
  const assetData = new Uint8Array(header.totalDataSize);
  let written = 0;
  let shocCount = 1;

  while (written < header.totalDataSize) {
    const topLevel = track.topLevelChunks[header.shocIndex + shocCount];
    if (!topLevel) break;
    shocCount++;

    if (topLevel.kind !== "SHOC") continue;

    const meta = topLevel.metadata;

    let chunkData: ArrayBufferSlice;
    if (meta.kind === "SDAT") {
      chunkData = meta.data;
    } else if (meta.kind === "Rdat") {
      chunkData = decompress(meta.data, meta.outBufferSize);
    } else {
      throw new Error("Unhandled SHOC type: " + meta.kind);
    }

    const size = Math.min(chunkData.byteLength, header.totalDataSize - written);
    assetData.set(chunkData.createTypedArray(Uint8Array, 0, size), written);
    written += size;
  }

  return new ArrayBufferSlice(assetData.buffer, 0, written);
}

export function getResourceList(track: TrackFile): RLst {
  const combined: RLst = { fileName: track.fileName, count: 0, entries: [] };
  const headers = getHeadersForType(track, "RLst");

  for (const header of headers) {
    const data = getDataForHeader(track, header);
    const rList = parseRLst(data, track.fileName);
    combined.count += rList.count;
    combined.entries.push(...rList.entries);
  }

  return combined;
}

export function getResource(
  track: TrackFile,
  resource: ResourceEntry,
): ParsedAsset {
  const header = getHeaderForResource(track, resource);
  if (!header)
    throw new Error(`Header not found for resource ${resource.resourceName}`);
  const data = getDataForHeader(track, header);

  switch (resource.typeTag) {
    case "Cact":
      return parseActor(data);
    case "obf ":
      return parseObf(data);
    case "o3d ":
      return parseO3D(false, data, header, resource.resourceName);
    case "o3da":
      return parseO3D(true, data, header, resource.resourceName);
    case "gmd ":
      return parseGmd(data, header, resource.resourceName);
    case "txf ":
    case "txf2": {
      const name = `${resource.resourceIndex}_${resource.resourceName}`;
      return parseTXF(data, header, name);
    }
    default:
      return parseGenericAsset(data, resource.typeTag.trimEnd(), header);
  }
}
